### PR TITLE
feat: Doesn't allow owners to exit a locked vehicle

### PR DESCRIPTION
**Pull request by Mesrine67**
addition of the check if the players are in the vehicle if yes lock on position 4 to lock the vehicle from the inside

### CODE REVIEW

**Review by solareon — CHANGES_REQUESTED** [review]
This seems counterintuitive to what most players expect with fivem vehicle keys. I gather that lock state 4 requires the vehicle to be unlocked before someone can get out aka like child safety locks on the back seat of most modern vehicles. Additionally we should be using `cache.vehicle` to check the vehicle of the player.

If we want to add something like this it should be configurable with the default off for those that want the "vanilla" experience.

File: client/main.lua [hunk 1]
```diff
@@ -87,7 +87,14 @@ local function setVehicleDoorLock(vehicle, state, anim)
...
                 lockstate = state and 2 or 1
             else
-                lockstate = (GetVehicleDoorLockStatus(vehicle) % 2) + 1 -- (1 % 2) + 1 -> 2  (2 % 2) + 1 -> 1 
+                local currentLockState = GetVehicleDoorLockStatus(vehicle)
+                local isPedInVehicle = IsPedInVehicle(cache.ped, vehicle, false)
+                -- Vérifier si le joueur est dans le véhicule et que l'état du verrou est 4, puis le passer à 1
+                if isPedInVehicle and currentLockState == 4 then
+                    lockstate = 1
+                else
+                    lockstate = isPedInVehicle and 4 or (currentLockState % 2) + 1 -- Utiliser l'état 4 si le joueur est à l'intérieur, sinon 1 ou 2
```
Comment (ArturMichalak) [inline client/main.lua:96]: Please use English in comments

**Review by ArturMichalak — CHANGES_REQUESTED** [review]

**Comment by Manason** [comment]
It's easier to review code if you commit the changes to the branch rather than ask us to review a comment. I think you are going in the right direction though. We generally use all lowercase for convars as well as store them in a local variable at the top of the file rather than retrieve via function.

File: client/main.lua [hunk 2]
```diff
@@ -367,15 +368,28 @@ end)
...
-    TriggerEvent('qb-vehiclekeys:client:ToggleEngine')
-end, false)
+local toggleLocksKeybind = lib.addKeybind({
+    name = 'togglelocks',
+    description = locale('info.toggle_locks'),
+    defaultKey = 'L',
+    onPressed = function(self)
+        local vehicle = getVehicle()
+        setVehicleDoorLock(vehicle, nil, true)
+        print(('pressed %s (%s)'):format(self.currentKey, self.name))
```
Comment (ArturMichalak) [inline client/main.lua:378]: Why are there prints here?

**Review by ArturMichalak — CHANGES_REQUESTED** [review]

File: client/main.lua (hunk 2 above)
Comment (ArturMichalak) [reply client/main.lua:378]: Check https://docs.qbox.re/release#code-quality

Comment (ArturMichalak) [reply client/main.lua:378]: However, use prints only when you really need them

File: client/main.lua [hunk 3]
```diff
@@ -66,33 +67,35 @@ local function getVehicle()
     return vehicle
 end
 
----manages the opening of locks
----@param vehicle number? The entity number of the vehicle.
----@param state boolean? State of the vehicle lock.
----@param anim any Aniation
+--- Manages the opening of locks
+--- @param vehicle number? The entity number of the vehicle.
```
Comment (ArturMichalak) [inline client/main.lua:71]: To be consistent with the project remove the white character between “---” and “@param”

File: client/main.lua [hunk 4]
```diff
@@ -367,15 +368,28 @@ end)
...
+        print(('released %s (%s)'):format(self.currentKey, self.name))
+    end
+})
+
+local engineKeybind = lib.addKeybind({
+    name = 'engine',
+    description = locale('info.engine'),
+    defaultKey = 'G',
+    onPressed = function(self)
+        TriggerEvent('qb-vehiclekeys:client:ToggleEngine')
```
Comment (ArturMichalak) [inline client/main.lua:390]: Consider using a function instead of an event.

**Review by ArturMichalak — CHANGES_REQUESTED** [review]

File: client/main.lua [hunk 5]
```diff
@@ -106,9 +109,7 @@ local function setVehicleDoorLock(vehicle, state, anim)
     else
         TriggerServerEvent('qb-vehiclekeys:server:setVehLockState', NetworkGetNetworkIdFromEntity(vehicle), 1)
     end
-end
-
-exports('SetVehicleDoorLock', setVehicleDoorLock)
+end exports('SetVehicleDoorLock', setVehicleDoorLock)
```
Comment (ArturMichalak) [inline client/main.lua:112]: keep the line breaking here

**Review by ArturMichalak — CHANGES_REQUESTED** [review]

**Comment by ArturMichalak** [comment]
Did you test this functionality?

File: client/main.lua (hunk 5 above)
Comment (Mesrine67) [reply client/main.lua:112]: sorry I'll review the code

**Comment by Mesrine67** [comment]
code tested ✅

File: client/main.lua [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 local config = require 'config.client'
 local functions = require 'client.functions'
+local childLockEnabled = GetConvar('qbx:vehiclekeys:enablechildlock', 'true') == 'true'
```
Comment (ArturMichalak) [inline client/main.lua:7]: Why should it be the default behavior?

File: client/main.lua [hunk 7]
```diff
@@ -66,33 +67,38 @@ local function getVehicle()
...
             else
-                lockstate = (GetVehicleDoorLockStatus(vehicle) % 2) + 1 -- (1 % 2) + 1 -> 2  (2 % 2) + 1 -> 1 
+                local currentLockState = GetVehicleDoorLockStatus(vehicle)
+                local isPedInVehicle = cache.vehicle == vehicle
+
+                if childLockEnabled and isPedInVehicle and currentLockState == 4 then
+                    -- Check if the player is in the vehicle and the lock state is 4, then switch it to 1
+                    lockstate = 1
+                else
+                    lockstate = isPedInVehicle and 4 or (currentLockState % 2) + 1 -- Use state 4 if the player is inside, otherwise 1 or 2
```
Comment (ArturMichalak) [inline client/main.lua:96]: Will someone be able to get in the car while someone is sitting in?

File: client/main.lua (hunk 6 above)
Comment (Mesrine67) [reply client/main.lua:7]: should we disable it by default? (I find that for the roleplay action the default setting true is best)

File: client/main.lua (hunk 7 above)
Comment (ArturMichalak) [reply client/main.lua:96]: What if the locks are open and someone tries to sit as a passenger?

Comment (ArturMichalak) [reply client/main.lua:96]: Of course, when the driver is already in the car.

Comment (Mesrine67) [reply client/main.lua:96]: mmmh thank you for making me think that I will adjust the code, I will adjust the code

File: client/main.lua (hunk 6 above)
Comment (ArturMichalak) [reply client/main.lua:7]: https://github.com/Qbox-project/qbx_vehiclekeys/pull/51#pullrequestreview-2081486256

**Comment by ArturMichalak** [comment]
Closing as author has not been responsive. Feel free to re-open.

**Pull request closed without merging**